File: skills/code-architect/references/code-examples/java/Builder.java
// Builder — construct complex objects step by step.

class Report {
    private final String title;
    private final String body;
    private final String footer;
    private final boolean numbered;

    private Report(Builder b) {
        this.title = b.title;
        this.body = b.body;
        this.footer = b.footer;
        this.numbered = b.numbered;
    }

    public String toString() {
        return "Report{title=" + title + ", body=" + body
                + ", footer=" + footer + ", numbered=" + numbered + "}";
    }

    static class Builder {
        private String title = "";
        private String body = "";
        private String footer = "";
        private boolean numbered = false;

        Builder title(String t) { this.title = t; return this; }
        Builder body(String b) { this.body = b; return this; }
        Builder footer(String f) { this.footer = f; return this; }
        Builder numbered(boolean n) { this.numbered = n; return this; }

        Report build() { return new Report(this); }
    }
}

public class Builder {
    public static void main(String[] args) {
        Report summary = new Report.Builder()
                .title("Quarterly")
                .body("Revenue up 12%")
                .numbered(true)
                .build();
        Report memo = new Report.Builder()
                .title("Memo")
                .footer("confidential")
                .build();
        System.out.println(summary);
        System.out.println(memo);
    }
}

File: skills/code-architect/references/code-examples/rust/builder.rs
// Builder — construct a complex object step by step, keeping construction logic
// separate from the final representation.

#[derive(Debug, Default)]
struct Report {
    title: String,
    sections: Vec<String>,
    footer: Option<String>,
}

#[derive(Default)]
struct ReportBuilder {
    report: Report,
}

impl ReportBuilder {
    fn new() -> Self {
        Self::default()
    }

    fn title(mut self, title: &str) -> Self {
        self.report.title = title.to_string();
        self
    }

    fn section(mut self, body: &str) -> Self {
        self.report.sections.push(body.to_string());
        self
    }

    fn footer(mut self, note: &str) -> Self {
        self.report.footer = Some(note.to_string());
        self
    }

    fn build(self) -> Report {
        self.report
    }
}

fn main() {
    let report = ReportBuilder::new()
        .title("Quarterly Summary")
        .section("Revenue is up.")
        .section("Costs are stable.")
        .footer("Confidential")
        .build();

    println!("# {}", report.title);
    for (i, s) in report.sections.iter().enumerate() {
        println!("  {}. {}", i + 1, s);
    }
    if let Some(f) = report.footer {
        println!("-- {} --", f);
    }
}
